Indent guides

bennetbo opened the pull request:
Builds on top of existing work from #2249, but here's a showcase:

https://github.com/zed-industries/zed/assets/53836821/4b346965-6654-496c-b379-75425d9b493f

TODO:
- [x] handle line wrapping
- [x] implement handling in multibuffer (crashes currently)
- [x] add configuration option
- [x] new theme properties? What colors to use?
- [x] Possibly support indents with different colors or background colors
- [x] investigate edge cases (e.g. indent guides and folds continue on empty lines even if the next indent is different)
- [x] add more tests (also test `find_active_indent_index`)
- [x] docs (will do in a follow up PR)
- [x] benchmark performance impact

Release Notes:

- Added indent guides ([#5373](https://github.com/zed-industries/zed/issues/5373))

bennetbo committed:
wip

bennetbo committed:
cleanup

bennetbo committed:
Merge branch 'main' into indent-guides-experiment

bennetbo committed:
wip line wrapping

d1y commented:
It looks great, I hope it can have built-in rainbow indentation :)

bennetbo committed:
support line wrapping

bennetbo committed:
adjust visible buffer range

bennetbo committed:
add indent guide colors to theme

bennetbo committed:
wip theme

bennetbo committed:
handle folded lines

bennetbo committed:
fix tests

bennetbo committed:
add accents to theme

Co-Authored-By: Nate Butler <1714999+iamnbutler@users.noreply.github.com>

bennetbo committed:
refactor settings

bennetbo committed:
theme adjustments

bennetbo committed:
handle fold correctly

bennetbo committed:
use correct offset for laying out indent guides

bennetbo committed:
wip multibuffer

bennetbo committed:
wip multibuffer active indent guides

bennetbo committed:
support active indent guide in multibuffer

bennetbo committed:
Merge branch 'main' into indent-guides-experiment

bennetbo committed:
handle empty lines correctly

bennetbo committed:
fix tests

bennetbo committed:
adjust buffer offset

bennetbo committed:
render indent guides that contain blocks correctly

bennetbo committed:
Merge branch 'main' into indent-guides-experiment

bennetbo committed:
truncate indent guides before excerpt headers

bennetbo committed:
wip active indent guide multibuffer

bennetbo committed:
simplify find_active_indent_guide_indices

bennetbo committed:
simplify code

bennetbo committed:
Merge branch 'main' into indent-guides-experiment

bennetbo committed:
add tests

bennetbo committed:
fix edge case

bennetbo committed:
Merge branch 'main' into indent-guides-experiment

bennetbo committed:
handle empty lines edge case

bennetbo committed:
rework active indentation

bennetbo committed:
fix folded indent guides

bennetbo committed:
optimize find_indented_row

bennetbo committed:
cleanup

bennetbo committed:
performance optimization

bennetbo committed:
Improve performance of finding the containing indent of a row

Co-Authored-By: Remco <djsmits12@gmail.com>

bennetbo committed:
Fixed edge cases in rope lines iterator

Co-Authored-By: Remco <djsmits12@gmail.com>

bennetbo committed:
Handle edge cases when computing enclosing indent

bennetbo committed:
Cleanup

bennetbo committed:
move indent guides to separate module

bennetbo committed:
compute active indent guides in background thread

bennetbo committed:
block with timeout implementation

bennetbo committed:
Merge branch 'main' into indent-guides-experiment

bennetbo committed:
fix merge

bennetbo committed:
cleanup

bennetbo committed:
clippy

bennetbo committed:
Do not access tab size for every row

bennetbo committed:
adjust setting names

bennetbo committed:
fix docs

bennetbo commented:
> It looks great, I hope it can have built-in rainbow indentation :)

@d1y It is actually built-in. In your settings file you will be able to do something like this:

```json
"experimental.theme_overrides": {
    "accents": [ // Specify some accent colors to use for the indent guide lines/background
      "#ffff40",
      "#7fff7f",
      "#ff7fff",
      "#4fecec",
      "#ffff40",
      "#7fff7f",
      "#ff7fff",
      "#4fecec",
      "#ffff40",
      "#7fff7f",
      "#ff7fff",
      "#4fecec"
    ]
  },
  "indent_guides": {
    "coloring": "indent_aware",  // Set this to indent_aware to paint the lines in rainbow colors
    "background_coloring": "indent_aware" // Set this to indent_aware to paint the background in rainbow colors
  },
```

You can also only colorize/disable the lines/the background by customizing the `coloring` and `background_coloring` values

<img width="772" alt="image" src="https://github.com/zed-industries/zed/assets/53836821/a31ff37a-e955-487d-98bf-3ceeafb3e30f">

bennetbo committed:
cleanup

bennetbo committed:
respect user accent overrides

baldwindavid commented:
@bennetbo Great work! For the "fixed" color mode is there a way to set the color of the active and inactive indent guide via the theme? I see an example for the gruvbox theme accents, but thinking that is only for "indent_aware" colors. What I would be shooting for is to set the "active" indent guide to a more prominent color and the inactive guides to transparent. Regardless of whether that is in here this will be a great addition. Thanks!

baldwindavid commented:
The other notes for future consideration are:
- It would be nice to be able to toggle on and off with a command like `editor::ToggleIndentGuides` (similar to `editor::ToggleLineNumbers`).
- Some languages tend need indent guides more than others. For example, I don't typically want guides in Elixir, but need them in HTML-ish code with deeper nesting. Might be nice to support language overrides similar to those for hard warp, tab size, etc.

I'm not even sure I'd want to mess with the default active/inactive theme colors mentioned in my previous comment if these sorts of settings were available.

SomeoneToIgnore commented on crates/language/src/buffer.rs:
```diff
@@ -3058,6 +3089,241 @@ impl BufferSnapshot {
+        result_vec
+    }
+
+    pub async fn enclosing_indent(&self, mut buffer_row: u32) -> Option<(Range<u32>, u32)> {
```
> u32, u32, 32

Passing by in curiosity, and wanted to note that using `MultiBufferRow` and `DisplayRow` might fix panics and reduce bugs.
Would be nice to remove all u32's possible around, but not insisting.

bennetbo commented:
@baldwindavid Indent guides use the following theme color's `editor.indent_guide` and `editor.indent_guide_active` when `coloring` is set to `fixed` (which is the default), so they can be styled by the theme and also customized with theme overrides.
Thanks for the suggestions! Will add them in a follow up PR

bennetbo committed:
store multi buffer rows in indent guides

bennetbo committed:
allow configuring indent guides per language

bennetbo committed:
clippy

bennetbo committed:
clippy

bennetbo committed:
clippy

bennetbo committed:
replace usages of u32 with row types

bennetbo committed:
disable indent guides in some editors

bennetbo committed:
fix default value for enable indent guide setting

bennetbo committed:
cleanup

bennetbo committed:
Merge branch 'main' into indent-guides-experiment

baldwindavid commented:
@bennetbo Super cool. I added `"editor.indent_guide": "#00000000"` and it works great.

I also see that you went ahead and added the per language config. I tried it out and quickly realized that my plan to disable for Elixir is kind of foiled in that this also disables it for heex (the elixir HTML template language). I tried this, but didn't seem to achieve what I was trying to do as the guides are still disabled for heex...

```json
  "languages": {
    "Elixir": {
      "language_servers": ["lexical", "!elixir-ls", "!next-ls", "..."],
      "indent_guides": {
        "enabled": false
      }
    },
    "HEEX": {
      "indent_guides": {
        "enabled": true
      }
    },
```

At any rate, don't think that has anything to do with this PR as the language-level settings do work. Thanks!

bennetbo commented:
> @bennetbo Super cool. I added `"editor.indent_guide": "#00000000"` and it works great.
> 
> I also see that you went ahead and added the per language config. I tried it out and quickly realized that my plan to disable for Elixir is kind of foiled in that this also disables it for heex (the elixir HTML template language). I tried this, but didn't seem to achieve what I was trying to do as the guides are still disabled for heex...
> 
> ```json
>   "languages": {
>     "Elixir": {
>       "language_servers": ["lexical", "!elixir-ls", "!next-ls", "..."],
>       "indent_guides": {
>         "enabled": false
>       }
>     },
>     "HEEX": {
>       "indent_guides": {
>         "enabled": true
>       }
>     },
> ```
> 
> At any rate, don't think that has anything to do with this PR as the language-level settings do work. Thanks!

I am not familiar with HEEX and elixir, is the HEEX code in the same file as the elixir code? Right now there's no support for rendering indent guides only for a certain part of a file (the language settings get inferred from the first line of the buffer, which in your case is probably elixir -> that's why they don't show up)

baldwindavid commented:
Elixir is the main language and HEEX a templating language for Elixir. They are both in the [elixir extension within the zed repo](https://github.com/zed-industries/zed/tree/main/extensions/elixir/languages).

While the elixir extension does support heex injections, the scenario I'm talking about here is for elixir in a file with a `.ex` extension and a heex template in a different file with the extension `.heex`.

bennetbo commented:
> Elixir is the main language and HEEX a templating language for Elixir. They are both in the [elixir extension within the zed repo](https://github.com/zed-industries/zed/tree/main/extensions/elixir/languages).
> 
> While the elixir extension does support heex injections, the scenario I'm talking about here is for elixir in a file with a `.ex` extension and a heex template in a different file with the extension `.heex`.

Hmm that is odd, I just tested it and specifying something like tab_size will also not work, HEEX always seems to use the settings that I specify for elixir
